ToolBox/KrakenDataDownloader/KrakenDataDownloader.cs

```diff
@@ -0,0 +1,166 @@
+        /// <param name="symbol">Symbol for the data we're looking for.</param>
+        /// <param name="resolution">Resolution of the data request</param>
+        /// <param name="startUtc">Start time of the data in UTC</param>
+        /// <param name="endUtc">End time of the data in UTC</param>
+        /// <returns>Enumerable of base data for this symbol</returns>
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            if (endUtc < startUtc)
+            {
+                throw new ArgumentException("The end date must be greater or equal than the start date.");
+            }
+
+            if (resolution != Resolution.Tick)
+            {
+                throw new NotSupportedException("Only Tick Resolution is supported.");
+            }
+
+            var startUnixTime = ToUnixTime(startUtc) * 1000000000; // Multiply by 10^9 per Kraken API
+            var endUnixTime = ToUnixTime(endUtc) * 1000000000;
+            var url = string.Format(UrlPrototype, symbol.Value, startUnixTime);
+            List<List<string>> data;
+
+            using (var client = new WebClient())
+            {
+                var response = client.DownloadString(url);
+                dynamic result = JsonConvert.DeserializeObject<dynamic>(response);
+                if (result.error.Count != 0)
+                {
+                    throw new Exception("Error in Kraken API: " + result.error[0]);
+                }
+
+                data = result.result[symbol.Value].ToObject<List<List<string>>>();
+
+                foreach (var i in data)
+                {
+                    var time = FromUnixTime(Convert.ToInt64(i[2].Split('.')[0]));
+                    if (time > endUtc)
+                    {
+                        break;
+                    }
+
+                    var value = Decimal.Parse(i[0]);
+                    var volume = Decimal.Parse(i[1]);
+
+                    yield return new Tick
+                    {
+                        Value = value,
+                        Time = time,
+                        DataType = MarketDataType.Tick,
+                        Symbol = symbol,
+                        TickType = TickType.Trade,
+                        Quantity = volume,
+                        Exchange = "kraken"
+                    };
+                }
+
+                var last = Convert.ToInt64(result.result.last);
+                while (last < endUnixTime)
+                {
+                    url = string.Format(UrlPrototype, symbol.Value, last);
+                    response = client.DownloadString(url);
+                    result = JsonConvert.DeserializeObject<dynamic>(response);
+
+                    var errorCount = 0;
+                    while (result.error.Count != 0 && errorCount < 10)
+                    {
+                        errorCount++;
+                        System.Threading.Thread.Sleep(6000);
```

Comment:
I'm guessing this is self-imposed rate limiting?

If so, there's a pretty easy to use type available, `QuantConnect.Util.RateGate`. You can use it like this:
```csharp
// define gate allowing, for example, 10 calls per 10 minutes -- which is different than 1 call per minute
var rateGate = new RateGate(10, TimeSpan.FromMinutes(10));
// blocking call until the gate allows you through
rateGate.WaitToProceed();
```